pq-crypto/s2n_kyber_512_evp.c

```diff
@@ -20,57 +20,63 @@
 
 #include "error/s2n_errno.h"
 #include "tls/s2n_kem.h"
+#include "utils/s2n_safety.h"
 #include "utils/s2n_safety_macros.h"
 
 #if defined(S2N_LIBCRYPTO_SUPPORTS_KYBER512)
-int s2n_kyber_512_evp_generate_keypair(uint8_t *public_key, uint8_t *private_key) {
-    EVP_PKEY_CTX *kyber_pkey_ctx = EVP_PKEY_CTX_new_id(EVP_PKEY_KYBER512, NULL);
+
+DEFINE_POINTER_CLEANUP_FUNC(EVP_PKEY *, EVP_PKEY_free);
+DEFINE_POINTER_CLEANUP_FUNC(EVP_PKEY_CTX *, EVP_PKEY_CTX_free);
+
+int s2n_kyber_512_evp_generate_keypair(uint8_t *public_key, uint8_t *secret_key) {
+    DEFER_CLEANUP(EVP_PKEY_CTX *kyber_pkey_ctx = EVP_PKEY_CTX_new_id(EVP_PKEY_KEM, NULL), EVP_PKEY_CTX_free_pointer);
     POSIX_GUARD_PTR(kyber_pkey_ctx);
+    POSIX_ENSURE(EVP_PKEY_CTX_kem_set_params(kyber_pkey_ctx, NID_KYBER512_R3), S2N_FAILURE);
     POSIX_ENSURE(EVP_PKEY_keygen_init(kyber_pkey_ctx), S2N_FAILURE);
 
-    EVP_PKEY *kyber_pkey = NULL;
+    DEFER_CLEANUP(EVP_PKEY *kyber_pkey = NULL, EVP_PKEY_free_pointer);
     POSIX_ENSURE(EVP_PKEY_keygen(kyber_pkey_ctx, &kyber_pkey), S2N_FAILURE);
 
     size_t public_key_size = S2N_KYBER_512_R3_PUBLIC_KEY_BYTES;
-    size_t private_key_size = S2N_KYBER_512_R3_SECRET_KEY_BYTES;
+    size_t secret_key_size = S2N_KYBER_512_R3_SECRET_KEY_BYTES;
     POSIX_ENSURE(EVP_PKEY_get_raw_public_key(kyber_pkey, public_key, &public_key_size), S2N_FAILURE);
-    POSIX_ENSURE(EVP_PKEY_get_raw_private_key(kyber_pkey, private_key, &private_key_size), S2N_FAILURE);
+    POSIX_ENSURE(EVP_PKEY_get_raw_private_key(kyber_pkey, secret_key, &secret_key_size), S2N_FAILURE);
 
     return S2N_SUCCESS;
 }
 
 int s2n_kyber_512_evp_encapsulate(uint8_t *ciphertext, uint8_t *shared_secret,
                                   const uint8_t *public_key) {
     size_t public_key_size = S2N_KYBER_512_R3_PUBLIC_KEY_BYTES;
-    EVP_PKEY *kyber_pkey = EVP_PKEY_new_raw_public_key(EVP_PKEY_KYBER512, NULL, public_key, public_key_size);
+    DEFER_CLEANUP(EVP_PKEY *kyber_pkey = EVP_PKEY_kem_new_raw_public_key(NID_KYBER512_R3, public_key, public_key_size), EVP_PKEY_free_pointer);
```

Comment:
Our CI should catch any memory leaks for code hit by the unit tests. Was this code previously never executed anywhere? Is it being executed now? If you remove the DEFER_CLEANUP, does the CI correctly fail on one of the ASAN or valgrind tests?

Reply:
> Was this code previously never executed anywhere? Is it being executed now?

the code was never previously executed in CI because S2N_AWSLC_KYBER_UNSTABLE was never specified in the CI scripts.

> Is it being executed now?

yes. this is ensured by the new unit test.

> If you remove the DEFER_CLEANUP, does the CI correctly fail on one of the ASAN or valgrind tests?

I can submit a test commit to try this out. will report back.

Reply:
DEFER_CLEANUP removal: https://github.com/WillChilds-Klein/s2n-tls/commit/8c5fb223e44e775e870fac793a1e1527d8ad1b05

build job: https://tiny.amazon.com/g0516bql/isenamazfede

Reply:
the removal of DEFER_CLEANUP appears to have resulted in the expected memory leak:

![Screen Shot 2023-03-09 at 6 03 11 PM](https://user-images.githubusercontent.com/3589880/224180138-b0f3ad01-0526-418a-999b-cb35ec44b3db.png)

Reply:
Nice, I'm convinced we've got this working now :)